### PR TITLE
new plugin: BetterSettings ~ improves Discord's settings

### PR DESCRIPTION
**Pull request by Kyuuhachi**
Makes your settings-menu-opening experience more pleasant by removing unnecessary delays and organizing the right-click menu.

Also cleans up a hack related to the right click menu in the core settings plugin.

### CODE REVIEW

**Comment by Oi-jojo** [comment]
arent these just a combination of your fast menu and setting subsection plugins?

**Comment by Kyuuhachi** [comment]
Yeah I saw little point in having them separate.

**Comment by Oi-jojo** [comment]
o alr i see, cant wait for this to get merged

File: src/plugins/betterSettings/index.tsx [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    description: "Enhances your settings-menu-opening experience",
+    authors: [Devs.Kyuuhachi],
+    settings,
+
+    patches: [
+        {
+            find: "this.renderArtisanalHack()",
+            replacement: [
+                { // Fade in on layer
+                    match: /(?<=(\w+)\.contextType=\w+\.AccessibilityPreferencesContext;)/,
```
Comment (Vendicated) [inline src/plugins/betterSettings/index.tsx:51]: please use \i

File: src/plugins/betterSettings/index.tsx [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            replacement: {
+                match: /(?<=handleOpenSettingsContextMenu.{0,250}?\i\.el\(("\d+")\)\.then.*?Messages\.USER_SETTINGS,)(?=onClick:)/,
+                replace: "onMouseEnter(){let r=Vencord.Webpack.wreq;r.el($1).then(r.bind(r,$1));},"
+            },
+            predicate: () => settings.store.eagerLoad,
+        },
+        { // Settings cog context menu
+            find: "Messages.USER_SETTINGS_ACTIONS_MENU_LABEL",
+            replacement: {
+                match: /\(0,\i.default\)\(\)(?=\.filter)/,
```
Comment (Vendicated) [inline src/plugins/betterSettings/index.tsx:87]: maybe this could be made a bit more specific

File: src/plugins/betterSettings/index.tsx [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                    replace: "\"div\""
+                },
+            ],
+            predicate: () => settings.store.disableFade,
+        },
+        { // Load menu stuff on hover, not on click
+            find: "Messages.USER_SETTINGS_WITH_BUILD_OVERRIDE.format",
+            replacement: {
+                match: /(?<=handleOpenSettingsContextMenu.{0,250}?\i\.el\(("\d+")\)\.then.*?Messages\.USER_SETTINGS,)(?=onClick:)/,
+                replace: "onMouseEnter(){let r=Vencord.Webpack.wreq;r.el($1).then(r.bind(r,$1));},"
```
Comment (Vendicated) [inline src/plugins/betterSettings/index.tsx:80]: this is potentially dangerous. what if discord also has a variable called r? i'd just remove the variable

File: src/plugins/betterSettings/index.tsx [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        { // For some reason standardSidebarView also has a small fade-in
+            find: "DefaultCustomContentScroller:function()",
+            replacement: [
+                {
+                    match: /\(0,\i\.useTransition\)\((\i)/,
+                    replace: "(_cb=>_cb(void 0,$1))||$&"
+                },
+                {
+                    match: /\i\.animated\.div/,
+                    replace: "\"div\""
```
Comment (Vendicated) [inline src/plugins/betterSettings/index.tsx:71]:
```suggestion
                    replace: '"div"'
```

File: src/plugins/betterSettings/index.tsx (hunk 3 above)
Comment (Kyuuhachi) [reply src/plugins/betterSettings/index.tsx]: I don't personally see an issue with it, but sure.

Comment (Vendicated) [reply src/plugins/betterSettings/index.tsx]: the issue is that discords code might contain a variable in their code named "r" which would conflict and crash discord

just avoid creating variables or name them something definitely unique like "vcFooBar"

File: src/plugins/betterSettings/index.tsx [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        default: true,
+        restartNeeded: true
+    },
+    organizeMenu: {
+        description: "Organizes the settings cog context menu",
+        type: OptionType.BOOLEAN,
+        default: true
+    },
+    eagerLoad: {
+        description: "Eagerly load menu contents",
```
Comment (Vendicated) [inline src/plugins/betterSettings/index.tsx:33]: what menu contents?

File: src/plugins/betterSettings/index.tsx [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+const settings = definePluginSettings({
+    disableFade: {
+        description: "Disable the crossfade animation",
+        type: OptionType.BOOLEAN,
+        default: true,
+        restartNeeded: true
+    },
+    organizeMenu: {
+        description: "Organizes the settings cog context menu",
```
Comment (Vendicated) [inline src/plugins/betterSettings/index.tsx:28]:
```suggestion
        description: "Organizes the settings cog context menu into categories",
```

File: src/plugins/betterSettings/index.tsx [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                replace: "$self.wrapMenu($&)"
+            }
+        }
+    ],
+
+    Layer({ mode, baseLayer = false, ...props }: {
+        mode: "SHOWN" | "HIDDEN";
+        baseLayer?: boolean;
+    } & HTMLAttributes<HTMLDivElement>) {
+        const hidden = mode === "HIDDEN";
```
Comment (Vendicated) [inline src/plugins/betterSettings/index.tsx:97]: currently if this component errors, it will fully brick Discord

Comment (Vendicated) [reply src/plugins/betterSettings/index.tsx]: fixed via latest commit

**Pull request merged**